### PR TITLE
Feature/save excel

### PR DESCRIPTION
**Pull request by netashefer**

### CODE REVIEW

File: src/db/db.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+
+import { FastifyInstance } from 'fastify';
+import fastifyPlugin from 'fastify-plugin';
+import { Client } from 'pg';
+
+const client = new Client({
```
Comment (YahelR) [inline src/db/db.ts:6]: can you add here a todo for me to research how to put this config without it being on the git? Not like it has any sensitive info, but I think if it doesn't take much I should change it to be more secure :)

File: src/db/queryExecuter.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+export const executeQuery = async (client: any, query: string) => {
```
Comment (YahelR) [inline src/db/queryExecuter.ts]: it's not query in our system though, it should be named accordingly

File: src/index.ts [hunk 3]
```diff
@@ -2,16 +2,18 @@ import cors from '@fastify/cors';
...
 fastify.register(rootRouter);
+fastify.register(dbConnector);
 
 const start = async () => {
     try {
         await fastify.listen({ port: PORT });
         console.log(`server is listening on port ${PORT}`);
     } catch (err) {
+        console.log(err);
         fastify.log.error(err);
```
Comment (YahelR) [inline src/index.ts]: wait why do we have two logs here? what is fastify.log.error? where does it go

File: src/routes/excel/excelRoute.ts [hunk 4]
```diff
@@ -1,14 +1,17 @@
...
+import { DataSource } from "../../types/excel.types";
 import { excelPostSchema } from "./schemas";
 
 const excelRoute = async (fastify: FastifyInstance) => {
 
-    fastify.post<{Body: {table: Table}}>('/parse', {schema: excelPostSchema}, async (request, reply) => {
-        const table = request.body.table;
-        const parsedTable = new ExcelParserService().parseTable(table);
-        return parsedTable;
+    fastify.post<{ Body: DataSource; }>('/add', { schema: excelPostSchema }, async (request, reply) => {
```
Comment (YahelR) [inline src/routes/excel/excelRoute.ts]: I feel like add is not quite descriptive enough, try thinking about a better route name but you can also just leave it be for the time being

File: src/routes/excel/excelRoute.ts [hunk 5]
```diff
@@ -1,14 +1,17 @@
...
+        const { table, displayName, dashboardId } = request.body;
+        const parsedTable = new ExcelService().parseTable(table);
+
+        // @ts-ignore - todo: make type of fastify.db.client
+        const dataSourceId = new ExcelService().addDataSource(fastify.db.client, { displayName, dashboardId, table: parsedTable });
```
Comment (YahelR) [inline src/routes/excel/excelRoute.ts]: no need for a const, just return it

File: src/schemas/models/excelFile.schema.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+export const excelDataSource: OpenAPIV3.SchemaObject = {
+    type: "object",
+    properties: {
+        table,
+        displayName: { type: "string" },
+        dashboardId: { type: "string" },
+    }
+};
+
+export const dataSourceId: OpenAPIV3.SchemaObject = {
```
Comment (YahelR) [inline src/schemas/models/excelFile.schema.ts:13]: move this to another file

File: src/services/excelService.ts [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+import { DATABASE_NAME } from "../db/db.constants";
+import { v4 as uuid } from 'uuid';
+const DIRECTORY_PATH = "C:/Users/neta1/Desktop/try1/";
+enum supportedDataTypes {
+    date = "(DATE)",
+    number = "(NUMBER)"
+}
+
+class ExcelService {
+    tableName = "dataSources";
```
Comment (YahelR) [inline src/services/excelService.ts:14]: why not use the enum of tablenames here

File: src/services/excelService.ts [hunk 8]
```diff
@@ -0,0 +1,65 @@
+import { ColumnParsingMap, DataSource } from "../types/excel.types";
+import { Table } from "../types/table.types";
+import * as XLSX from 'xlsx';
+import fs from 'fs';
+import { parse } from 'date-fns';
+import { executeQuery } from "../db/queryExecuter";
+import { DATABASE_NAME } from "../db/db.constants";
+import { v4 as uuid } from 'uuid';
+const DIRECTORY_PATH = "C:/Users/neta1/Desktop/try1/";
+enum supportedDataTypes {
```
Comment (YahelR) [inline src/services/excelService.ts:8]: move to another file

File: src/services/excelService.ts [hunk 9]
```diff
@@ -0,0 +1,65 @@
+import { ColumnParsingMap, DataSource } from "../types/excel.types";
+import { Table } from "../types/table.types";
+import * as XLSX from 'xlsx';
+import fs from 'fs';
+import { parse } from 'date-fns';
+import { executeQuery } from "../db/queryExecuter";
+import { DATABASE_NAME } from "../db/db.constants";
+import { v4 as uuid } from 'uuid';
```
Comment (YahelR) [inline src/services/excelService.ts:2]: import order

File: src/services/excelService.ts [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+            });
+            data.push(dataRow);
+        });
+
+        table.data = data;
+        return table;
+    }
+
+    saveExcelFile(file: XLSX.WorkBook) {
+        const dataSourceId = "filename";
```
Comment (YahelR) [inline src/services/excelService.ts]: move to a const that would be in another file :)
maybe excel.constants.ts

File: src/services/excelService.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        if (!fs.existsSync(DIRECTORY_PATH)) {
+            fs.mkdirSync(DIRECTORY_PATH, { recursive: true });
```
Comment (YahelR) [inline src/services/excelService.ts]: thats also cool!

File: src/services/excelService.ts [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+        }
+        XLSX.writeFileXLSX(file, `${DIRECTORY_PATH}${dataSourceId}.xlsx`);
```
Comment (YahelR) [inline src/services/excelService.ts]: so for now do we still save the excels in the server?

File: src/services/excelService.ts [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+        INSERT INTO ${DATABASE_NAME}."${this.tableName}"
+        ("dataSourceId", "displayName", "dashboardId", "dataTable")
+        VALUES ('${id}', '${excelDataSource.displayName}', '${excelDataSource.dashboardId}', '${JSON.stringify(excelDataSource.table)}')
+        ;`;
+        await executeQuery(client, query);
+        return id;
+    }
+}
+
+export default ExcelService;
```
Comment (YahelR) [inline src/services/excelService.ts]: make this a singleton

File: src/db/db.ts (hunk 1 above)
Comment (netashefer) [reply src/db/db.ts:6]: I saw places that use ENV variables, so it will be one way, I will add  TODO

File: src/db/queryExecuter.ts (hunk 2 above)
Comment (netashefer) [reply src/db/queryExecuter.ts]: we dont have query anyway, we call it dataSourceId, but I will change

File: src/index.ts (hunk 3 above)
Comment (netashefer) [reply src/index.ts]: it came like this, I will delete it

File: src/services/excelService.ts (hunk 10 above)
Comment (netashefer) [reply src/services/excelService.ts]: I removed the function, this was the problem witth saving. I forgot to delete it

**Review by YahelR — APPROVED** [review]

**Pull request merged**